Bluetooth: Classic: add local name

chengkai15 opened the pull request:
add BR/EDR local name

jhedberg reviewed (changes requested):
We already have `CONFIG_BT_DEVICE_NAME` and `bt_set_name()` APIs. I don't think it makes sense to do BR/EDR-specific APIs for this.

chengkai15 committed:
Bluetooth: Classic: move bt local name to bt_br_write_local_name

Extract the HCI Write Local Name command into a reusable
bt_br_write_local_name() helper and use it in bt_br_init().

When CONFIG_BT_DEVICE_NAME_DYNAMIC and CONFIG_BT_SETTINGS are
both enabled, skip writing the compile-time default here since
the settings commit callback will push the correct name (either
restored from flash or the default) to the controller.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Classic: sync local name to BR/EDR on settings commit

When CONFIG_BT_DEVICE_NAME_DYNAMIC is enabled, write the device name
(restored from flash or the compile-time default) to the BR/EDR
controller during commit_settings().

Replace the bt_set_name() call with a direct strncpy into bt_dev.name
to avoid an unnecessary flash write, then explicitly push the name to
the BR/EDR controller via bt_br_write_local_name().

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Classic: sync local name to BR/EDR in bt_set_name()

When CONFIG_BT_CLASSIC is enabled and the controller is ready,
push the updated device name to the BR/EDR controller via
bt_br_write_local_name() inside bt_set_name(). This ensures
runtime name changes are properly synchronized to the BR/EDR
transport.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -858,19 +858,13 @@ int bt_br_init(void)
 	}
 
+	strncpy(bt_dev.name, CONFIG_BT_DEVICE_NAME, CONFIG_BT_DEVICE_NAME_MAX);
+
```
I do not think so.

> int bt_enable(bt_ready_cb_t cb) { ...... } else if (IS_ENABLED(CONFIG_BT_DEVICE_NAME_DYNAMIC)) { **err = bt_set_name(CONFIG_BT_DEVICE_NAME);** if (err) { LOG_WRN("Failed to set device name (%d)", err); } } ...... }

If the settings is not enabled and CONFIG_BT_DEVICE_NAME_DYNAMIC is enabled, the name is stored into `bt_dev.name`. The default value is CONFIG_BT_DEVICE_NAME since the settings is not enabled.

> static int commit_settings(void) { ...... #if defined(CONFIG_BT_DEVICE_NAME_DYNAMIC) **if (bt_dev.name[0] == '\0') { bt_set_name(CONFIG_BT_DEVICE_NAME); }** #endif ......

Or if the settings is enabled and CONFIG_BT_DEVICE_NAME_DYNAMIC is enabled, the `bt_dev.name` is loaded from settings. If the name is not stored, the default value CONFIG_BT_DEVICE_NAME  will be used and store it to settings.

gzh-terry commented on subsys/bluetooth/host/classic/br.c:
`strlcpy`

chengkai15 replied:
strlcpy was not supported in zephyr

jhedberg reviewed (changes requested):
Meant to request changes along with my previous comments

chengkai15 commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4752,6 +4752,10 @@ int bt_set_name(const char *name)
 
+	if (len == 0) {
+		return -EINVAL;
+	}
```
it was removed. the original change was made to consider that the controller might not support empty character names, and the spec did not specifically indicate that this situation was not supported, so the original design was maintained.

chengkai15 committed:
Bluetooth: host: call bt_br_init() for Classic init in hci_init

When the controller supports BR/EDR and CONFIG_BT_CLASSIC is
enabled, call bt_br_init() to perform the full BR/EDR Classic
initialization (SSP mode, inquiry mode, local name, Class of
Device, etc.).

For non-Classic builds that still need BR/EDR buffer sizes,
gate the hci_read_buffer_size() fallback with
IS_ENABLED(CONFIG_BT_CONN) to make the build-time condition
explicit at the call site.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 commented:
> Meant to request changes along with my previous comments

done

lylezhu2012 reviewed (changes requested):
I think there is a case missing in this PR.
The device name has been stored into NVM. And re-initialize the Bluetooth Host Stack. 
The device name will not be written to the controller.

https://github.com/zephyrproject-rtos/zephyr/blob/ad7beb530a05388748e2c307ac4176dd139eb7ef/subsys/bluetooth/host/settings.c#L183-L196

lylezhu2012 commented on subsys/bluetooth/host/hci_core.c:
Did you double confirm?

The function bt_br_init() of file subsys/bluetooth/host/classic/br.c is not called anymore in your PR.

chengkai15 replied:
> Did you double confirm?
> 
> The function bt_br_init() of file subsys/bluetooth/host/classic/br.c is not called anymore in your PR.

Yes,  it  only fix the warning  that bt_br_init  was conflict with another global variable br.h, 

and I had to #include<classic/br.h> header because I was going to use  IS_ENABLED CONFIG_BT_CLASSIC condition
from @jhedberg ‘s suggestion

	**if (IS_ENABLED(CONFIG_BT_CLASSIC)) {**
		err = bt_br_write_local_name(bt_dev.name);
		if (err) {
			LOG_WRN("Unable to set local name");
			return err;
		}
	}

lylezhu2012 replied:
> name would only be get by RNR(remote name request), and EIR name feature would be add in later patch.

I do not think so. With the changes of PR, the initialisation of Bluetooth Classic will be passed (the function `bt_br_init()` will be called anymore). It means in your PR, a function that was working properly was changed to not work anymore. I think it is a regression.


> Yes, it only fix the warning that bt_br_init was conflict with another global variable br.h,
> 
> and I had to #include<classic/br.h> header because I was going to use IS_ENABLED CONFIG_BT_CLASSIC condition from @jhedberg ‘s suggestion

I understand why you change the function name. But I think he was just trying to give you some advice on the issue itself. I don't think he expects to solve the issue by introducing new issues.

I have some suggestions for you,
- Put the implementation of function `bt_br_write_local_name()` to file `hci_core.c`,
- Or, declare the prototype of the function `bt_br_write_local_name()` in file `hci_core.h`,
- Or, create a new internal header file for Classic, just likes as `br_internal.h` or `br_interface.h`.

lylezhu2012 commented on subsys/bluetooth/host/settings.c:
```diff
@@ -255,8 +255,13 @@ static int commit_settings(void)
 	if (bt_dev.name[0] == '\0') {
 		bt_set_name(CONFIG_BT_DEVICE_NAME);
+	} else {
+		bt_set_name(bt_dev.name);
```
There is a potential issue that the life of Flash will be reduced. Since there is not any change of device name, I think it is better to write device name to controller directly.

lylezhu2012 replied:
> When I initiate the `bt_set_name `method, my log shows that `commit_settings `is not triggered. Is this because my environment is not right?

The `commit_settings()` would be called when the function `settings_load()` called.

> do you means that if the name has been set before, and the **same name** would set again? 

Uh-huh, that is what your change is doing.

> `bt_set_name ` would check and will not take effect if it is set again

In theory I think this can be done. But currently, the new name is not checked in function `bt_set_name()`.

jhedberg commented:
@chengkai15 are you still working on this? There seems to be open review/discussion threads.

chengkai15 commented:
> @chengkai15 are you still working on this? There seems to be open review/discussion threads.

I'm still continuing, I've been delayed for some time

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4314,7 +4309,11 @@ static int hci_init(void)
-		err = bt_br_init();
+		if (IS_ENABLED(CONFIG_BT_CLASSIC)) {
+			err = bt_br_init();
+		} else {
```
If you have #ifdefs in the static function or where you're calling the static function, then both places need them. As long as the calls are not protected by #ifdef (e.g. around the entire function from where the calls happen) then `IS_ENABLED()` is possible.

chengkai15 commented on subsys/bluetooth/host/settings.c:
```diff
@@ -320,7 +321,25 @@ static int commit_settings(void)
+	 */
+	if (IS_ENABLED(CONFIG_BT_CLASSIC)) {
+		err = bt_br_write_local_name(bt_dev.name);
+		if (err) {
```
updated
